Zend/zend_execute.c

```diff
@@ -5012,6 +5012,13 @@ static zend_always_inline zend_result _zend_quick_get_constant(
 	}
 
 	if (!c) {
+		if (UNEXPECTED(
+			zend_string_equals_literal_ci(Z_STR_P(key), "exit")
+			|| zend_string_equals_literal_ci(Z_STR_P(key), "die")
+		)) {
+			zend_throw_unwind_exit();
+			return FAILURE;
+		}
```

Comment:
Could we achieve similar results with by de-sugaring `exit;` to a function call at the AST level? (e.g. `zend_compile_const()` can replace the AST with a function call AST and delegate to `zend_compile_call()`). This would remove the need for special cases at runtime, and also unify the behavior of `exit;` and `exit();` (wrt to disable_functions).

Reply:
I agree. Special-casing exit/die constant access doesn't sound like a cleaner solution than what we have now. Keeping them in the lexer/parser also keeps BC for AST-based tools.

Reply:
Yes, and it has side effects like defined(), constant() calls etc.

We may opt to special-case only statement-level "exit" and "die" constant fetches directly (i.e. handle in zend_compile_stmt, in the branch for constants).

Reply:
Oh, that's actually a smarter way of doing it!

I'll see if I can implement this, as I still struggle with the compiler part.

Reply:
> We may opt to special-case only statement-level "exit" and "die" constant fetches directly (i.e. handle in zend_compile_stmt, in the branch for constants).

I would prefer keeping parser support, as otherwise PHPStan and the likes need to adjust. I don't think there's a reason to break those.

Reply:
I was about to suggest something similar, because `exit;` not actually calling the function but just throwing that unwind exception probably leads to observable behavioral differences in debuggers dependent on whether the user wrote `exit;` or `exit();`, no?

Reply:
I don't think we should constrain ourselves with respect to phpstan etc.; there are worse BC breaks to be had.
I think the lexer, parser and opcodes are aside from any compatibilty guarantees.

Reply:
@ondrejmirtes has told me that this shouldn't be an issue if PHPParser still works, I'm asking @dseguy if this would affect Exakat or not.